### PR TITLE
Miscellaneous minor refactoring of the unit formatters

### PR DESCRIPTION
**Pull request by embray**
Most of these changes were driven by the observation I made, upon going to make some changes for #3986, that on some of the formatter classes (including the `Base` class) the `to_string` method was a normal instance method, while on others (about half) it was a `classmethod`.  This incongruity bugged me.

After looking over the other classes I decided it would be possible to make the `parse` and `to_string` methods on all the formatter classes into `classmethods`, so that now each formatter class is basically acting as a singleton (it would actually be a good use case for a general Singleton type, but I decided to leave that as an exercise for another time).

I think there are probably others bits of repetition between the formatter implementations, but I think that's about all I feel like addressing for now.

This also removed all inline imports from `parse` and `to_string` methods.  A bit of a hack, but doable.  Between eliminating formatter instantiation and inline imports, the time to instantiate a `Unit` from a string with the generic formatter (after already using the formatter at least once, so parser/lexer setup is excluded from this timing) went from about 13.1µs to about 9.2µs, or about 30%.  Not super-dramatic, but still worthwhile.

### CODE REVIEW

**Comment by embray** [comment]
Getting a bunch of mysterious file not closed errors that didn't occur before...

**Comment by mdboom** [comment]
That lazy classproperty is a nice addition and clearly makes writing singletons much easier.  These have always been singletons in concept, if not strictly speaking -- I was using instantiation as a trigger to generate the parsers the first time, but clearly these lazy classproperties are much nicer.  Be sure to put that in CHANGES as a separate item under utils -- I think that's one of the nice side effects of this PR.

**Comment by embray** [comment]
@mdboom The classproperty decorator is new as of a couple PRs ago anyways, and are already mentioned in the changelog.  I just added the "lazy" feature after realizing the use for this change.

Does still need a changelog entry to emphasize the change in formatters.

**Comment by embray** [comment]
Also glad they were always intended as singletons--I was trying to think if maybe there could be some case for some kind of instantiated formatter, but couldn't think of anything immediately.

**Comment by mdboom** [comment]
> Does still need a changelog entry to emphasize the change in formatters.

It's a minor point.  I've always considered any thing in `unit/format` private API.  (The public API is `unit.to_string(format=...)` etc.)

File: astropy/units/format/generic.py [hunk 1]
```diff
@@ -58,41 +57,50 @@ class Generic(Base):
...
+        'SIGN',
+        'UINT',
+        'UFLOAT'
+    )
+
+    @classproperty(lazy=True)
+    def _all_units(cls):
+        return cls._generate_unit_names()
+
+    @classproperty(lazy=True)
```
Comment (mhvk) [inline astropy/units/format/generic.py]: These ones don't really have to be lazy at all, right?  I'd think only `_all_units` would need to be calculated only once.

File: astropy/utils/decorators.py [hunk 2]
```diff
@@ -308,6 +308,24 @@ class classproperty(property):
...
+        The function that computes the value of this property (in particular,
+        the function when this is used as a decorator) a la `property`.
+
+    doc : str, optional
+        The docstring for the property--by default inherited from the getter
+        function.
+
+    lazy : bool, optional
+        If True, caches the value returned by the first call to the getter
+        function, so that it is only caused once (used for lazy evaluation
```
Comment (mhvk) [inline astropy/utils/decorators.py]: caused -> calculated

Comment (mhvk) [inline astropy/utils/decorators.py]: or "called"

**Comment by mhvk** [comment]
@embray - nice indeed!

File: astropy/units/format/generic.py (hunk 1 above)
Comment (embray) [inline astropy/units/format/generic.py]: No, not strictly necessary.  It doesn't matter much either way in this case.

File: astropy/utils/decorators.py (hunk 2 above)
Comment (embray) [inline astropy/utils/decorators.py]: Ah, nice catch.

**Comment by embray** [comment]
For some reason this wasn't re-picked up by CI watchers when I rebased.  Trying again....

**Pull request merged**